pkg/cloud/cloud.go

```diff
@@ -684,6 +680,61 @@ func (c *cloud) DeleteDisk(ctx context.Context, volumeID string) (bool, error) {
 	return true, nil
 }
 
+// executes a batched DescribeInstances API call
+func execBatchDescribeInstances(svc ec2iface.EC2API, input []string) (map[string]*ec2.Instance, error) {
+	klog.V(7).InfoS("execBatchDescribeInstances", "instanceIds", input)
+	request := &ec2.DescribeInstancesInput{
+		InstanceIds: aws.StringSlice(input),
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), batchDescribeTimeout)
+	defer cancel()
+
+	resp, err := describeInstances(ctx, svc, request)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make(map[string]*ec2.Instance)
+
+	for _, instance := range resp {
+		if instance.InstanceId == nil {
+			klog.Warningf("execBatchDescribeInstances: skipping instance: %v, reason: missing volume ID", instance)
+			continue
+		}
+		result[*instance.InstanceId] = instance
+	}
+
+	klog.V(7).InfoS("execBatchDescribeInstances: success", "result", result)
+	return result, nil
+}
+
+// batchDescribeInstances processes a DescribeInstances request by queuing the task and waiting for the result.
+func (c *cloud) batchDescribeInstances(request *ec2.DescribeInstancesInput) (*ec2.Instance, error) {
+	var task string
+
+	if len(request.InstanceIds) == 1 && request.InstanceIds[0] != nil {
+		task = *request.InstanceIds[0]
+	} else {
+		return nil, fmt.Errorf("batchDescribeInstances: invalid request, request: %v", request)
+	}
+
+	ch := make(chan batcher.BatchResult[*ec2.Instance])
+
+	b := c.bm.instanceIDBatcher
+	b.AddTask(task, ch)
+
+	r := <-ch
+
+	if r.Err != nil {
+		return nil, r.Err
+	}
+	if r.Result == nil {
+		return nil, fmt.Errorf("batchDescribeInstances: no instance found %s", task) // TODO Q: Should this be Cloud's ErrNotFound instead?
```

Comment:
> TODO Q: Should this be Cloud's ErrNotFound instead?

Yes +1 to this, that error type is more appropriate

Reply:
I will change `batchDescribeVolumes` as well then to be consistent.